lib/srv/app/aws/endpoints.go

```diff
@@ -128,6 +152,130 @@ func resolveEndpointByXForwardedHost(r *http.Request, headerKey string) (*endpoi
 	}, nil
 }
 
+// shouldDiscoverTimestreamEndpoint returns true if Timestream endpoint
+// discovery is required.
+//
+// Common Timestream operations require an "endpoint discovery" prior to the
+// actual API call. The "endpoint discovery" is done by making a
+// "DescribeEndpoints" call using the user credentials. However, in Endpoint
+// mode (e.g. `--endpoint-url` for AWS CLI or `Endpoint=` for JDBC), the
+// "endpoint discovery" is skipped, and the endpoint passed to the client is
+// expected to be an URL equivalent to what "DescribeEndpoints" returns. Thus
+// the app agent has to do the "endpoint discovery" first (using the user
+// credentials) to figure out where to send the actual APIs.
+//
+// Sample AWS SDK reference for the discovery flow:
+// https://github.com/aws/aws-sdk-go/blob/41717ba2c04d3fd03f94d09ea984a10899574935/service/timestreamquery/api.go#L1295-L1319
+func shouldDiscoverTimestreamEndpoint(r *http.Request) bool {
+	target := r.Header.Get("X-Amz-Target")
+	return strings.HasPrefix(target, timestreamOpPrefix) && !strings.HasSuffix(target, "DescribeEndpoints")
+}
+
+// resolveTimestreamEndpoint performs a Timestream endpoint discover to resolve
+// the endpoint.
+func (s *signerHandler) resolveTimestreamEndpoint(r *http.Request) (*endpoints.ResolvedEndpoint, error) {
+	sessCtx, err := common.GetSessionContext(r)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	awsAuthHeader, err := awsutils.ParseSigV4(r.Header.Get(awsutils.AuthorizationHeader))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	credentials, err := s.CredentialsGetter.Get(s.closeContext, awsutils.GetCredentialsRequest{
+		Provider:    s.Session,
+		Expiry:      sessCtx.Identity.Expires,
+		SessionName: sessCtx.Identity.Username,
+		RoleARN:     sessCtx.Identity.RouteToApp.AWSRoleARN,
+		ExternalID:  sessCtx.App.GetAWSExternalID(),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key := resolveTimestreamEndpointCacheKey{
+		credentials:              credentials,
+		region:                   awsAuthHeader.Region,
+		isTimestreamWriteRequest: isTimestreamWriteRequest(r),
+	}
+	re, err := utils.FnCacheGet(s.closeContext, s.cache, key, func(ctx context.Context) (*endpoints.ResolvedEndpoint, error) {
```

Comment:
We only cache it for a minute in `FnCache` which is the minimum of `CachePeriodInMinutes` (tho it seems `CachePeriodInMinutes` is usually 1440m (24h) by default). Note that since the endpoint can be different depending on who calls, I am using the user sessions' credentials here as part of the cache key. I can add a comment on the TTL here.

Seems a lot of work is required to get `crr.EndpointCache` working. And it only supports `string` key, and the `crr.Endpoint` is not thread-safe. So personally I would prefer to keep it simple with `FnCache`.

Reply:
I'm +1 to using AWS SDK's endpoint cache if we can, acknowledging the issue with thread safety. The `Discoverer` looks pretty minor to implement -- does the thread safety of `EndpointCache` matter here given that the Cache itself seems to be thread safe?

Reply:
I haven't tried an example with race detector. so this is speculation by inspecting this code
https://github.com/aws/aws-sdk-go/blob/5fabc200455abcde45b34541ee557eb910ab0bd6/aws/crr/cache.go#L28-L42

The map is a sync.Map but `ev.Prune()` is not concurrent safe. But I guess two concurrent `ev.Prune` probably end up the same thing.

I don't mind giving it a try or maybe implementing our own. but let's say the API returns 24h. Do we want to cache it for that long?

Reply:
regarding thread safety: I think it should be fine because the endpoint cache is storing/loading endpoint structs by value, so the endpoint returned from the sync.Map load will be a copy.

timestreamquery has an implementation of discover here which you could pretty much copy https://github.com/aws/aws-sdk-go/blob/5fabc200455abcde45b34541ee557eb910ab0bd6/service/timestreamquery/api.go#L493-L538

>  Note that since the endpoint can be different depending on who calls

I understand that this would be the case for different AWS accounts, but is it true that the endpoint can differ depending on caller within the same account?

> but let's say the API returns 24h. Do we want to cache it for that long?

The endpoint cache is an LRU cache so you can control the size of it. Looks like AWS opted to just use 10 for the size in their usage.

My main concern was that we might use an endpoint after it had expired. If the minimum TTL is 1 minute then that isn't an issue, although we will probably make a lot more endpoint discovery api calls than necessary. I leave it up to you to decide, so I won't block the PR anymore

Reply:
> We do support multiple accounts through AssumeRole. I guess we can extract Account ID from the role ARN if we want.

Oh, right 👍 

> Seems the endpoint cache is not LRU but it deletes a random entry.

You're right I think, the godoc for it seems to be inaccurate then.

Not sure what to pick for a cache size, but you can limit the number of endpoint addresses you add to each `crr.Endpoint` (as the code is doing now by just picking the first endpoint address)